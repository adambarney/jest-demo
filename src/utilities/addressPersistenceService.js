import axios from 'axios'

export function saveAddress(address) {
    //post address to webservice using axios
    //return webservice response, containing all addresses in address book

  return axios.post("example.com/address", address)
    .then(response => response.data)
}