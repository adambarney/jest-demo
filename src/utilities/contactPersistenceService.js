import axios from 'axios'

function saveContact(contact) {
  //POST contact to webservice using axios
  //return webservice response, containing all contacts in DB
}





























// export function saveContact(contact) {
//   //POST contact to webservice using axios
//   //return webservice response, containing all contacts in DB
//
//   return axios.post('http://localhost:8080/contact', contact)
//     .then(response => response.data)
// }

function getContacts() {
  return axios.get('http://localhost:8080/contact')
    .then(response => response.data)
}

export {
  saveContact,
  getContacts
}
