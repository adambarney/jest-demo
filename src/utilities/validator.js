export function validatePhone(phoneNumber) {
  if (phoneNumber.length === 10) {
    document.getElementById('submit-button').disabled = false
    return true
  } else {
    document.getElementById('submit-button').disabled = true
    return false
  }
}






































// export function validatePhone(phoneNumber) {
//   if (phoneNumber.length === 10) {
//     document.getElementById('save-button').disabled = false
//     return true
//   } else {
//     document.getElementById('save-button').disabled = true
//     return false
//   }
// }


