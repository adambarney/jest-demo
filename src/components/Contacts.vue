<template>
  <div>
    <v-container>
      <h1>Contacts</h1>
      <v-row>
        <v-col cols="3">
          <v-form>
            <v-text-field
                data-test="name-input"
                label="Name"
                v-model="contactName">
            </v-text-field>
            <v-text-field
                data-test="phone-input"
                label="Phone Number"
                v-model="phoneNumber">
            </v-text-field>
            <v-btn data-test="submit-button"
                   id="submit-button"
                   @click="saveNewContact">Save
            </v-btn>
          </v-form>
        </v-col>
        <v-col>
          <v-data-table data-test="contact-table"
              :headers=headers
              :items=allContacts></v-data-table>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script>
import {getContacts, saveContact} from '../utilities/contactPersistenceService'

export default {
  name: 'Contacts',
  data: function () {
    return {
      contactName: '',
      phoneNumber: '',
      headers: [
        {text: 'Name', value: 'name'},
        {text: 'Phone Number', value: 'phoneNumber'}
      ],
      allContacts: []
    }
  },
  mounted() {
    this.getExistingContacts()
  },
  methods: {
    saveNewContact: function () {
      saveContact({
        name: this.contactName,
        phoneNumber: this.phoneNumber
      }).then((response) => {
        this.allContacts = response
      })
    },
    getExistingContacts: function () {
      getContacts().then((response) => {
        this.allContacts = response
      })
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.container {
  max-width: 960px;
}

</style>
