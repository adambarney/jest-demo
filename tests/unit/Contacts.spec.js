import {createLocalVue, mount} from '@vue/test-utils'
import flushPromises from 'flush-promises'
import Vuetify from 'vuetify'
import Contacts from '../../src/components/Contacts'
import * as contactService from '../../src/utilities/contactPersistenceService'

const localVue = createLocalVue()
const vuetify = new Vuetify()
let wrapper

const existingContact = {
  name: 'Joe',
  phoneNumber: '5131234567'
}
const newContact = {
  name: 'Jenny',
  phoneNumber: '5138675309'
}
const expectedContacts = [existingContact, newContact]

beforeEach(async () => {
  contactService.getContacts = jest.fn()
    .mockImplementation(() => {
      return Promise.resolve([existingContact])
    })
  contactService.saveContact = jest.fn()
    .mockImplementation(() => Promise.resolve(expectedContacts))

  wrapper = mount(Contacts, {
    localVue,
    vuetify
  })
  await flushPromises()
})



describe('Save button', () => {
  it('should save new contact when clicked, and update contacts table (v.1)', async () => {
    wrapper.find('[data-test=name-input]').setValue('Jenny')
    wrapper.find('[data-test=phone-input]').setValue('5138675309')
    wrapper.find('[data-test=submit-button]').trigger('click')
    await flushPromises()
    expect(wrapper.find('[data-test=contact-table]').find('tr:nth-child(2)').text()).toEqual('Jenny5138675309')
    expect(wrapper.find('[data-test=contact-table]').element.textContent).toContain('Joe5131234567')
  })

  it('should save new contact when clicked, and update contacts table (v.2)', async () => {
    wrapper.vm.contactName = 'Jenny'
    wrapper.vm.phoneNumber = '5138675309'
    wrapper.vm.saveNewContact()
    await flushPromises()
    expect(wrapper.vm.allContacts).toEqual(expectedContacts)
  })
})


describe('contacts', () => {
  it('should render correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})