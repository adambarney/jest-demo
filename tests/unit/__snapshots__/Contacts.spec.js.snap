// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contacts should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <h1>
      Contacts
    </h1>
     
    <div
      class="row"
    >
      <div
        class="col col-3"
      >
        <form
          class="v-form"
          novalidate="novalidate"
        >
          <div
            class="v-input theme--light v-text-field"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-3"
                    style="left: 0px; position: absolute;"
                  >
                    Name
                  </label>
                  <input
                    data-test="name-input"
                    id="input-3"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="v-input theme--light v-text-field"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-6"
                    style="left: 0px; position: absolute;"
                  >
                    Phone Number
                  </label>
                  <input
                    data-test="phone-input"
                    id="input-6"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <button
            class="v-btn v-btn--contained theme--light v-size--default"
            data-test="submit-button"
            id="submit-button"
            type="button"
          >
            <span
              class="v-btn__content"
            >
              Save
          
            </span>
          </button>
        </form>
      </div>
       
      <div
        class="col"
      >
        <div
          class="v-data-table v-data-table--has-bottom theme--light"
          data-test="contact-table"
        >
          <div
            class="v-data-table__wrapper"
          >
            <table>
              <colgroup>
                <col
                  class=""
                />
                <col
                  class=""
                />
              </colgroup>
              <thead
                class="v-data-table-header"
              >
                <tr>
                  <th
                    aria-label="Name: Not sorted. Activate to sort ascending."
                    aria-sort="none"
                    class="text-start sortable"
                    role="columnheader"
                    scope="col"
                  >
                    <span>
                      Name
                    </span>
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                      style="font-size: 18px;"
                    />
                  </th>
                  <th
                    aria-label="Phone Number: Not sorted. Activate to sort ascending."
                    aria-sort="none"
                    class="text-start sortable"
                    role="columnheader"
                    scope="col"
                  >
                    <span>
                      Phone Number
                    </span>
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                      style="font-size: 18px;"
                    />
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  class=""
                >
                  <td
                    class="text-start"
                  >
                    Joe
                  </td>
                  <td
                    class="text-start"
                  >
                    5131234567
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div
            class="v-data-footer"
          >
            <div
              class="v-data-footer__select"
            >
              Rows per page:
              <div
                class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select"
              >
                <div
                  class="v-input__control"
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-owns="list-17"
                    class="v-input__slot"
                    role="button"
                  >
                    <div
                      class="v-select__slot"
                    >
                      <div
                        class="v-select__selections"
                      >
                        <div
                          class="v-select__selection v-select__selection--comma"
                        >
                          10
                        </div>
                        <input
                          aria-label="Rows per page:"
                          aria-readonly="false"
                          autocomplete="off"
                          id="input-17"
                          readonly="readonly"
                          type="text"
                        />
                      </div>
                      <div
                        class="v-input__append-inner"
                      >
                        <div
                          class="v-input__icon v-input__icon--append"
                        >
                          <i
                            aria-hidden="true"
                            class="v-icon notranslate mdi mdi-menu-down theme--light"
                          />
                        </div>
                      </div>
                      <input
                        type="hidden"
                        value="10"
                      />
                    </div>
                    <div
                      class="v-menu"
                    >
                      <!---->
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="v-data-footer__pagination"
            >
              1-1 of 1
            </div>
            <div
              class="v-data-footer__icons-before"
            >
              <button
                aria-label="Previous page"
                class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
                disabled="disabled"
                type="button"
              >
                <span
                  class="v-btn__content"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate mdi mdi-chevron-left theme--light"
                  />
                </span>
              </button>
            </div>
            <div
              class="v-data-footer__icons-after"
            >
              <button
                aria-label="Next page"
                class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
                disabled="disabled"
                type="button"
              >
                <span
                  class="v-btn__content"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate mdi mdi-chevron-right theme--light"
                  />
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
